src/Juvix/Backends/Michelson/Compilation/Term.hs

```diff
@@ -7,167 +7,276 @@ import Juvix.Backends.Michelson.Compilation.Prim
+          insts ← evaluateAndPushArgs arguments primTy args paramTy
+          recurseApplication ([], arguments {- the arguments to the primitive -}, ann {- will be inlined -}) primTy args insts paramTy
+        t → do
+          failWith ("Applications applied to non lambda term: " <> show t)
+
+-- captures, arguments, body are for the function
+-- lamTy is the type of the function
+-- args are the arguments to which it is being applied
+recurseApplication ∷
+  ∀ m a.
+  ( HasState "stack" VStack.T m,
+    HasThrow "compilationError" CompilationError m,
+    HasWriter "compilationLog" [CompilationLog] m
+  ) ⇒
+  ([Symbol], [Symbol], Term) →
+  J.Type PrimTy PrimVal →
+  [a] →
+  [Op] →
+  M.Type →
+  m (Either LamPartial M.ExpandedOp)
+recurseApplication (captures, lamArguments, body) lamTy args insts paramTy = do
+  let argsL = length args
+      lamArgsL = length lamArguments
+  case compare argsL lamArgsL of
+    EQ → do
+      -- Exactly applied case, we have the correct number of arguments.
+      -- need to assert that we correctly drop them later, this is complicated!
+      -- need to pass things to be dropped along with the partially applied function datatype
+      eitherFuncOp ← termToInstr body paramTy
+      case eitherFuncOp of
+        Right f → do
+          -- Fully applied case without a lambda: evaluate all the arguments, inline the function, drop the args.
+          pure $
+            Right
+              ( M.SeqEx
+                  ( insts
+                      <> [ f,
+                           M.PrimEx
+                             (M.DIP (replicate argsL (M.PrimEx M.DROP)))
+                         ]
+                  )
+              )
+        Left (LamPartial ops captures remArgs body ty) → do
+          -- Fully applied case with a returned lambda: return the instructions followed by the lambda, defer evaluation of the body.
+          -- Will this type be correct if we compile this to a lambda?
+          pure (Left (LamPartial (insts <> ops) captures remArgs body ty))
+    LT → do
+      -- Underapplied case, we don't yet have enough arguments to inline.
+      -- Here we are turning formerly bound arguments into captures.
+      let (evaluatedArgs, remainingArgs) = splitAt argsL lamArguments
+      lamTy ← dropNArgs lamTy argsL
+      pure (Left (LamPartial insts (captures <> evaluatedArgs) remainingArgs body lamTy))
+    GT → do
+      -- Overapplied case, we must figure out what the body is and recurse.
+      let (args, extraApplied) = splitAt lamArgsL args
+      eitherFuncOp ← termToInstr body paramTy
+      case eitherFuncOp of
+        Right _ → failWith "invalid case"
+        Left (LamPartial ops parCaptures parArgs parBody parTy) → do
+          -- todo: are these captures correct
+          -- Will this type be correct if we compile this to a lambda?
+          recurseApplication (parCaptures, parArgs, parBody) parTy extraApplied (insts <> ops) paramTy
```

Comment:
Will we correctly drop the initially evaluated arguments (`insts`) here (later on)? I'm not sure that we will

Reply:
I don't see where we drop any of this, correct